### DISCUSSION TITLE
Enabling <form> functionality on customElements

### DISCUSSION THREAD

**Discussion by ctrain149** [lit/lit#2204] [Ideas]
I love LitElement but I find one of the major drawbacks of using Lit is the requirement of the ShadowDOM. The ShadowDOM is a double-edged sword in practice, because it gives us boundaries between customElements and prevents low-level communication between customElements too. In this discussion I want to take a look at ```<form>``` and autocomplete.

The requirements needed to enable autocomplete on your <input> elements are:
1. We need our ```<input>``` elements wrapped in a ```<form>``` tag
2. We need an ```<input type="submit">``` included within the ```<form>``` tag
3. We need to give our ```<input>``` elements the autocomplete attribute with a proper value

There is an issue, in practice, for LitElement. If we design/import customElements that wrap around an ```<input>``` tag then we have created a boundary around our ```<input>``` using the ShadowDOM. If we try to use the elements like this...

``` javascript
<form>
  <custom-input-component autocomplete="email">
    <-- ShadowDom -->
    <input autocomplete="${this.autocomplete}" />
  </custom-input-component>

  <custom-input-component autocomplete="current-password">
    <-- ShadowDom -->
    <input autocomplete="${this.autocomplete}"/>
  </custom-input-component>

  <input type="submit">
</form>
```

...then we find that autocomplete does NOT work here. The ShadowDOM is preventing the ```<form>``` from seeing the ```<input>``` component here which breaks our requirements. I was hoping, maybe, with the revival of Lit that we could see some Tender Loving Care for those developers who would like to use both of these functionalities? I know this is difficult...

**Comment by e111077** [comment]
Note: our excellent tech writer has written an article on web.dev on this. You can read it [here](https://web.dev/more-capable-form-controls/).

## Autocomplete

For autocomplete, that seems to be a bug with Chrome. You can follow along here:

https://bugs.chromium.org/p/chromium/issues/detail?id=649162

## Form Participation

For element participation there are two specs still being implemented by the browser vendors:

1. [ElementInternals (Form Participation API)](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals)
2. [FormDataEvent.set](https://developer.mozilla.org/en-US/docs/Web/API/FormDataEvent)

### ElementInternals

Allows the browser to see your custom element as a form element and will do everything you expect it to do.

Problem is that it's hard to polyfill and thus the polyfills are quite large:

* [Polyfill](https://www.npmjs.com/package/element-internals-polyfill)
* [External resources](https://css-tricks.com/creating-custom-form-controls-with-elementinternals/)

### FormDataEvent.set

Allows your component to hook into a form rather than the other way around. Your element needs to find the form in its ancestor tree and then listen for the `submit` event and manually add its field to the event. It's also super easy to polyfill.

Problem is that you won't get the nice things about form submission like the sections autofilling with previous data on back button press. Also a lot of things are not just _automatic_ like `ElementInternals`.

* [Polyfill (seems like it's not published?)](https://github.com/webcomponents/polyfills/tree/master/packages/formdata-event)
* [Polyfill bug](https://github.com/webcomponents/polyfills/issues/172)
* [External resources (original spec proposal)](https://docs.google.com/document/d/1JO8puctCSpW-ZYGU8lF-h4FWRIDQNDVexzHoOQ2iQmY/edit)

### Workarounds

You can easily polyfill the formdata event. The issue linked above even has a few-lined example on how to do so.

Alternatively you can disable shadow dom for those components by overriding `createRenderRoot() {/*decide how to handle styles*/;return this;}`.

Another, older method was to create another, hidden, input in the light dom of the component and mirror the value, name, and type of the input inside the shadow DOM.

Finally, another popular method is to write your component such that the input is slotted in by the user. e.g.

```html
<my-datepicker>
  <input slot="month" name="birthday-month">
  <input slot="day" name="birthday-day">
  <input slot="year" name="birthday-year">
</my-datepicker>
```
